k-distribution/include/kframework/builtin/domains.md

```diff
@@ -1055,6 +1055,8 @@ module INT-SYMBOLIC [symbolic]
   rule X modInt N => X requires 0 <=Int X andBool X <Int N [simplification]
   rule X   %Int N => X requires 0 <=Int X andBool X <Int N [simplification]
 
+  rule _ ^Int N => 0 requires N <Int 0 [simplification]
+
```

Comment:
I'd like to see a test that exercises this new Z3 encoding. Anything that sends the symbol to Z3 on the Haskell backend should do, because it was causing crashes previously.

```
requires "domains.md"

module TEST
    imports BOOL
    imports INT-SYMBOLIC

    syntax KItem ::= runLemma ( Bool ) | doneLemma ( Bool )
 // -------------------------------------------------------
    rule <k> runLemma(B) => doneLemma(B) ... </k>

    rule N ^Int N' <=Int M ^Int M' => true
      requires 0 <Int N andBool 0 <Int N'
       andBool 0 <Int M andBool 0 <Int M'
       andBool N <=Int M andBool N' <=Int M'
      [simplification, smt-lemma]
                     
endmodule
                     
module TEST-SPEC
    imports TEST
                     
    claim <k> runLemma(3 ^Int X <=Int 3 ^Int Y) => doneLemma(true) ... </k>
      requires 0 <Int X andBool 0 <Int Y
       andBool X <=Int Y
endmodule
```

I'd also like Daejun to double-check this new Z3 encoding of `_^Int_`, to make sure we are getting the correct semantics.

Reply:
Ping @daejunpark

Reply:
@ehildenb @Baltoli  The simplification rule, `N < 0  \implies  X ^ N == 0`, doesn't seem correct, when X is 1 or -1.   When X = 1, it is 1.  When X = -1, it is alternating between 1 (if N is even) and -1 (if N is odd).

Reply:
@ehildenb Note that it is not only for the simplification rule, but also for the z3 encoding in the hook: `(ite (< #2 0) 0 (to_int (^ #1 #2)))`.   As I mentioned, this encoding is incorrect for `#1` being 1 or -1.

Reply:
So do we want to outright remove the simplification rule, and add another check to the SMT hook that handles the cases for 1 / -1?

Reply:
Well, I think we could also change the smt hook to say `(and (< 0 #1) (< 0 #2))`, so that we don't have to think about the 0/negative cases. Then we could add K rules for specifically the various 0 cases `N ^ 0 requires 0 <Int N`, `0 ^ N requires 0 <Int N`, `0 ^ 0`.

Reply:
Let me clarify a bit, to make sure we are all on the same page:

Let `X` and `N` be an integer.  Let `Y` be `\trunc(X ^ N)`.   (`\trunc` means the truncation rounding, e.g., `\trunc(-0.1) = 0`, `\trunc(0.1) = 0`, `\trunc(1.2) = 1`, etc.)

Then we have:
1. `X = 0` implies `( if N > 0, then Y = 0, else ( if N < 0, then Y = \infty, else Y = ??? ) )`   (see below for `0^0`)
2. `X = 1` implies `Y = 1`
3. `X = -1` implies  `( if N is even, then Y = 1, else Y = -1 )`
4. `(X > 1 or X < -1) and N < 0` implies `Y = 0`

Note that:
- `N = 0`  does _not_ necessarily imply  `Y = 1`  until we define `0^0 = 1`.
- `N < 0`  does _not_ imply  `Y = 0`.   (See (1), (2), and (3) above for the counter-examples.)


Regarding `0^0`, many mathematicians seem to agree on `0^0 = 1` (see [here](https://www.maa.org/book/export/html/116806)), while WolframAlpha considers it [undefined](https://www.wolframalpha.com/input/?i=0%5E0).  We need to double-check what's the interpretation of z3 for that.

Reply:
My simple suggestion would be to just put a guard for `to_int`, like: `(implies (>= #2 0) (to_int (^ #1 #2)))`.

Those who want to reason about the negative exponent case, would need to come up with their own lemmas, simplifications, and/or custom z3 encodings.

Reply:
FWIW, I think that if the type system contains both `+\infty` and `-\infty`, then `0^N` should be undefined when `N < 0`

Reply:
Just adding one point here: as mentioned by @daejunpark, we should double-check the interpretation of `0 ^ 0`, and, if I'm not mistaken, ```**``` is underlying implementation of ```^``` for z3, correct? I've made a script to test `x=0`, `y=0`, `n=x**y` using different versions of z3's python module. Bellow is its output (version followed by the generated model, if can be achieved):
```

4.4.2.1.post1: [n = 0**0, y = 0, x = 0]
4.5.1.0: [n = 0**0, y = 0, x = 0]
4.5.1.0.post1: [n = 0**0, y = 0, x = 0]
4.5.1.0.post2: [n = 0**0, y = 0, x = 0]
4.8.0.0.post1: [n = 0**0, y = 0, x = 0]
4.8.5.0: [n = 0**0, y = 0, x = 0]
4.8.6.0: [n = 0**0, y = 0, x = 0]
4.8.7.0: [n = 0, y = 0, x = 0]
4.8.8.0: [n = 0, y = 0, x = 0]
4.8.9.0: [n = 0, y = 0, x = 0]
4.8.10.0: unknown
4.8.11.0: unknown
4.8.12.0: unknown
```

In this case, we may end up getting different outcomes for different versions of z3, so I think a more safe approach would be something on the lines of what @ehildenb said (adding K rules for the specific 0 and negative cases).